### PR TITLE
Kate - LA101.ag_FAO_R_C_Y.R

### PR DESCRIPTION
**Pull request by kvcalvin**
Includes documentation, testing flags, etc.

### CODE REVIEW

File: R/zchunk_LA101.ag_FAO_R_C_Y.R [hunk 1]
```diff
@@ -40,79 +44,131 @@ module_aglu_LA101.ag_FAO_R_C_Y_DISABLED <- function(command, ...) {
...
-    #   ... new code with a fix
-    # }
-    #
-    #
-    # NOTE: there are `merge` and/or 'match' calls in this code. Be careful!
-    # For more information, see https://github.com/JGCRI/gcamdata/wiki/Merge-and-Match
-    # ===================================================
-
+    # Process FAO food consumption data (tons): remove unnecessary columns, convert units, aggregate to region and commodity
+    L100.FAO_ag_Food_t %>% select( -countries, -country.codes, -item.codes, -element, -element.codes ) %>%           # Remove unnecessary columns
```
Comment (bpbond) [inline R/zchunk_LA101.ag_FAO_R_C_Y.R]: Put `select` on its own line

File: R/zchunk_LA101.ag_FAO_R_C_Y.R [hunk 2]
```diff
@@ -40,79 +44,131 @@ module_aglu_LA101.ag_FAO_R_C_Y_DISABLED <- function(command, ...) {
...
-      add_comments("can be multiple lines") %>%
+    L101.ag_Food_Mt_R_C_Y %>%
+      add_title("FAO food consumption by GCAM region, commodity, and year") %>%
+      add_units("Mt/yr") %>%
+      add_comments("Aggregates FAO data by GCAM region, commodity, and year") %>%
+      add_comments("Data is also converted from tons to Mt") %>%
       add_legacy_name("L101.ag_Food_Mt_R_C_Y") %>%
-      add_precursors("precursor1", "precursor2", "etc") %>%
+      add_precursors("L100.FAO_ag_Food_t", "aglu/FAO_ag_items_cal_SUA", "common/iso_GCAM_regID") %>%
       # typical flags, but there are others--see `constants.R`
```
Comment (bpbond) [inline R/zchunk_LA101.ag_FAO_R_C_Y.R]: Probably remove this line and ones like it below

File: R/zchunk_LA101.ag_FAO_R_C_Y.R [hunk 3]
```diff
@@ -40,79 +44,131 @@ module_aglu_LA101.ag_FAO_R_C_Y_DISABLED <- function(command, ...) {
...
+      filter( GCAM_commodity != "NA" ) %>%                                                                           # Remove commodities not included in GCAM
+      mutate( value = value * Mcal_t * CONV_MCAL_PCAL )  %>%                                                         # Convert from tons to Pcal
+      group_by( GCAM_region_ID, GCAM_commodity, year ) %>%                                                           # Group by region, commodity, year
+      summarize( value=sum( value ) ) %>%                                                                            # Aggregate then map to appropriate data frame
+      ungroup() %>%                                                                                                  # Ungroup before complete
+      complete( GCAM_region_ID = unique( iso_GCAM_regID$GCAM_region_ID ),
+                GCAM_commodity, year, fill = list( value = 0 ) ) ->  L101.ag_Food_Pcal_R_C_Y                         # Fill in missing region/commodity combinations with 0
+    
+    #Calculate average caloric content of consumed commodities (kcal/g)
```
Comment (bpbond) [inline R/zchunk_LA101.ag_FAO_R_C_Y.R]: Space after `#`

File: R/zchunk_LA101.ag_FAO_R_C_Y.R [hunk 4]
```diff
@@ -40,79 +44,131 @@ module_aglu_LA101.ag_FAO_R_C_Y_DISABLED <- function(command, ...) {
...
+      left_join( FAO_ag_items_cal_SUA, by=c( "item" ) )  %>%                                                         # Map in GCAM commodities
+      filter( GCAM_commodity != "NA" ) %>%                                                                           # Remove commodities not included in GCAM
+      mutate( value = value * CONV_TON_MEGATON )  %>%                                                                # Convert from tons to Mt
+      group_by( GCAM_region_ID, GCAM_commodity, year ) %>%                                                           # Group by region, commodity, year
+      summarize( value=sum( value ) ) %>%                                                                            # Aggregate then map to appropriate data frame
+      ungroup() %>%                                                                                                  # Ungroup before complete
+      complete( GCAM_region_ID = unique( iso_GCAM_regID$GCAM_region_ID ),
+                GCAM_commodity, year, fill = list( value = 0 ) ) -> L101.ag_Food_Mt_R_C_Y                            # Fill in missing region/commodity combinations with 0
```
Comment (bpbond) [inline R/zchunk_LA101.ag_FAO_R_C_Y.R]: I'd suggest assignments on their own line, e.g.

```R
  GCAM_commodity, year, fill = list( value = 0 ) ) -> 
  L101.ag_Food_Mt_R_C_Y
```

File: R/zchunk_LA101.ag_FAO_R_C_Y.R [hunk 5]
```diff
@@ -40,79 +44,131 @@ module_aglu_LA101.ag_FAO_R_C_Y_DISABLED <- function(command, ...) {
...
-    # }
-    #
-    #
-    # NOTE: there are `merge` and/or 'match' calls in this code. Be careful!
-    # For more information, see https://github.com/JGCRI/gcamdata/wiki/Merge-and-Match
-    # ===================================================
-
+    # Process FAO food consumption data (tons): remove unnecessary columns, convert units, aggregate to region and commodity
+    L100.FAO_ag_Food_t %>% select( -countries, -country.codes, -item.codes, -element, -element.codes ) %>%           # Remove unnecessary columns
+      left_join_error_no_match( iso_GCAM_regID, by=c( "iso") )  %>%                                                  # Map in ISO codes
```
Comment (bpbond) [inline R/zchunk_LA101.ag_FAO_R_C_Y.R]: Do you mind doing a find-and-replace to remove spaces after open, and before close, parentheses? I.e. not do GCAM-style?

**Comment by kvcalvin** [comment]
I made all requested changes.

File: R/zchunk_LA101.ag_FAO_R_C_Y.R [hunk 6]
```diff
@@ -40,79 +44,141 @@ module_aglu_LA101.ag_FAO_R_C_Y_DISABLED <- function(command, ...) {
...
+      summarize(value=sum(value)) %>%                                                                              # Aggregate then map to appropriate data frame
+      ungroup() %>%                                                                                                # Ungroup before complete
+      complete(GCAM_region_ID = unique(iso_GCAM_regID$GCAM_region_ID),
+                GCAM_commodity, year, fill = list(value = 0)) ->                                                   # Fill in missing region/commodity combinations with 0
+      L101.ag_Food_Pcal_R_C_Y
+    
+    # Calculate average caloric content of consumed commodities (kcal/g)
+    L101.ag_Food_Pcal_R_C_Y %>% 
+      left_join(L101.ag_Food_Mt_R_C_Y, by = c("GCAM_region_ID", "GCAM_commodity", "year")) %>%                  # Join food in Mt to food in Pcal
+      mutate(value = ifelse(value.y == 0, 1, value.x / value.y))  %>%                                           # Calculate average caloric content, set NA values to 1
```
Comment (bpbond) [inline R/zchunk_LA101.ag_FAO_R_C_Y.R]: @kvcalvin I'll make this change but FYI - you *never* want to use `ifelse` in a pipeline like this; use `if_else` (the dplyr version) instead.

File: R/zchunk_LA101.ag_FAO_R_C_Y.R [hunk 7]
```diff
@@ -40,79 +44,141 @@ module_aglu_LA101.ag_FAO_R_C_Y_DISABLED <- function(command, ...) {
...
-    # NOTE: there are `merge` and/or 'match' calls in this code. Be careful!
-    # For more information, see https://github.com/JGCRI/gcamdata/wiki/Merge-and-Match
-    # ===================================================
-
+    # Process FAO food consumption data (tons): remove unnecessary columns, convert units, aggregate to region and commodity
+    L100.FAO_ag_Food_t %>% 
+      select(-countries, -country.codes, -item.codes, -element, -element.codes) %>%                                # Remove unnecessary columns
+      left_join_error_no_match(iso_GCAM_regID, by=c("iso"))  %>%                                                   # Map in ISO codes
+      left_join(FAO_ag_items_cal_SUA, by=c("item"))  %>%                                                           # Map in GCAM commodities
+      filter(GCAM_commodity != "NA") %>%                                                                           # Remove commodities not included in GCAM
```
Comment (bpbond) [inline R/zchunk_LA101.ag_FAO_R_C_Y.R]: I will change, but FYI should be `filter(!is.na(GCAM_commodity)) %>%`

(NA isn't a string.)

**Comment by bpbond** [comment]
Wow, what a complete PITA this was. @kvcalvin I understand why you were so frustrated. 

I couldn't get the `L101.ag_HA_bm2_R_C_Y` and `L101.ag_Prod_Mt_R_C_Y` outputs to pass tests, even though I could confirm that they're holding the same data. I'm pretty sure this is because the old data system writes data out using `write.csv` while we're using `readr::write_csv` and subtle differences creep in, probably more so as one progresses through the data system.

The best thing I could think of was to create a less-strict test, that compares the sums of all numeric columns (as opposed to using `all.equal`). Your code now tags these two datasets (and only these two) with a flag to use this test. Not perfect, but everything now passes, and it should still give us a warning if things change.

Take a look at the update I'm about to sync, and if things seem good, I'll merge.

**Pull request merged**